### ISSUE TITLE
torch shape error when training probing models

### ISSUE THREAD

**Issue by rmovva** [nelson-liu/contextual-repr-analysis#7]
I'm trying to train tagging probes using precomputed representations from the encoder of my Transformer-based machine translation model, but am getting the following error when training attempts to begin:

```
2020-07-29 05:23:27,936 - INFO - allennlp.training.trainer - Training
...
  File "/raj-learn/contextual-repr-analysis/contexteval/models/tagger.py", line 340, in forward
    average=self.loss_average)                                                                                                                             
  File "/raj-learn/envs/contextual_repr_analysis/lib/python3.6/site-packages/allennlp/nn/util.py", line 633, in sequence_cross_entropy_with_logits
    negative_log_likelihood_flat = - torch.gather(log_probs_flat, dim=1, index=targets_flat)
RuntimeError: invalid argument 2: Input tensor must have same size as output tensor apart from the specified dimension at /pytorch/aten/src/THC/generic/THCTensorScatterGather.cu:29
[INFO/MainProcess] process shutting down  
```

I've computed these representations and saved them in a correctly-formatted hdf5 file using the `make_hdf5_file` function from the README. I then confirmed that this hdf5 file looks as expected: the keys are string id's that map to (N_tokens, 1024) numpy arrays. N_tokens is the length of that sentence and 1024 is my encoder representation dim (which I correctly pass into my config file). sentence_to_index also looks fine, mapping newline-stripped sentences to string id's.

I get this error when training begins with both the CCG supertagging task and the PTB POS tagging task (which both use the WSJ sentences dataset).

Here's my config file for CCG supertagging (modeled off of yours for ELMo original):
```
{
    "dataset_reader": {
        "type": "ccg_supertagging",
        "contextualizer": {
            "type": "precomputed_contextualizer",
            "representations_path": "../data/precomputed_reps/ccg_wsj_sentences_all.hdf5"
        }
    },
    "validation_dataset_reader": {
        "type": "ccg_supertagging",
        "contextualizer": {
            "type": "precomputed_contextualizer",
            "representations_path": "../data/precomputed_reps/ccg_wsj_sentences_all.hdf5"
        }
    },
    "train_data_path": "../data/probing_task_data/ccg/train.txt",
    "validation_data_path": "../data/probing_task_data/ccg/dev.txt",
    "test_data_path": "../data/probing_task_data/ccg/test.txt",
    "evaluate_on_test" : true,
    "model": {
        "type": "tagger",
        "token_representation_dim": 1024
    },
    "iterator": {
        "type": "basic",
        "batch_size" : 80
    },
    "trainer": {
        "num_epochs": 50,
        "patience": 3,
        "cuda_device": 0,
        "validation_metric": "+accuracy",
        "optimizer": {
            "type": "adam",
            "lr": 0.001
        }
    }
}
```

I'm going to try looking into this error now and will update if I figure out what's going on. I cloned the contextual-repr-analysis repo and installed dependencies from requirements.txt, so I imagine I should have the version of allennlp + conllu that were used at the time this package was written.

Thanks again for your help with this excellent work!

**Comment by nelson-liu** [comment]
Hm, I'd try adding a breakpoint at https://github.com/nelson-liu/contextual-repr-analysis/blob/master/contexteval/models/tagger.py#L339 and verifying that the length of logits is indeed equal to the length of the labels. How are you calculating `n_tokens`?

**Comment by rmovva** [comment]
Ok, it indeed looks like the shape of logits ([80, 55, 45] for batch 1) doesn't match the length of labels ([80, 51]). The logits are longer, so I guess that means my sentence tokenizer (Moses) differs from the tokenizer used by the PTB POS tagging, which I didn't think to check beforehand. Do you know which tokenization scheme is used to generate labels for the PTB POS tags? EDIT: Sorry, I just saw the README says tokens are space delimited.

Sadly I don't think changing my model's tokenizer is very trivial, since I trained with Moses, so I can't just swap something else in easily. But ideally once I figure out how Moses differs from the PTB tagging tokenizer, I can map my tokens onto the labelled tokens in some structured way.

**Comment by nelson-liu** [comment]
> Do you know which tokenization scheme is used to generate labels for the PTB POS tags? EDIT: Sorry, I just saw the README says tokens are space delimited.

Yeah, most tagging datasets are already pre-tokenized.

> Sadly I don't think changing my model's tokenizer is very trivial, since I trained with Moses, so I can't just swap something else in easily.

Tokenization is the worst :( sorry about that. For BPE-based models (that also differ in tokenization), I just set each PTB token's embedding to be its last BPE subword token. Maybe a similar strategy could be applied here...

**Comment by rmovva** [comment]
I ended up being able to do something similar. Basically I take my Moses-tokenized output (which mostly splits more things than the tagging datasets split), and then match my tokens back to the tagging dataset's tokens. Most of the differences are explained by possessives (Moses splits "Emma 's" --> "Emma ' s") and contractions ("does n't" --> "n 't") so I was able to rescue a large fraction of the dataset. 

Then, in this repo, I had to change the POS-specific data reader function to allow for KeyErrors in the sentence_to_index dict, and skip sentences that we weren't able to get representations for. Because the data reading function is implemented differently for each probing task, I guess I'll have to make that change for each task, but that's not a big deal.

Thanks for the suggestion!
